zigzag-circuit.md

```diff
@@ -0,0 +1,179 @@
+
+#### Private Inputs
+
+*Inputs that the prover uses to generate a SNARK proof, these are not needed by the verifier to verify the proof*
+
+- `CommR : [LAYERS-1]Fr`: Commitment of the the encoded data at each layer. 
+
+  Note: Size is `LAYERS-1` since the commitment to the last layer is `CommRLast`
+
+- `DataProof : [LAYERS][][TREE_DEPTH-2]Fr`: Merkle tree inclusion proof for the current layer unencoded challenged nodes.
+
+  Note: Size of proof per layer is `TREE_DEPTH-2` because it excludes (1) the roothash, and (2) the leaf value: (1) the root hash of the proof will be the corresponding unencoded data commitment for the current layer `l` (either `CommD` or `CommR[l-1]`), (2) the leaf value is `DataValue[l][c]`
+
+- `ReplicaProof : [LAYERS][][TREE_DEPTH-2]Fr`: Merkle tree inclusion proof for the current layer encoded challenged nodes.
+
+  Note: Size of proof per layer is `TREE_DEPTH-2` because it excludes (1) the roothash, and (2) the leaf value: (1) the root hash of the proof will be the corresponding unencoded data commitment for the current layer  (either `CommR[l]` or `CommRLast`), (2) the leaf value is `ReplicaValue[l][c]`
+
+- `ParentProof : [LAYERS][][PARENT_NODES][TREE_DEPTH-2]Fr`: Pedersen hashes of the Merkle inclusion proofs of the parent nodes for each challenged node at layer `l`.
+
+  Note: Size of proof per layer is `TREE_DEPTH-2` because it excludes (1) the roothash, and (2) the leaf value: (1) the root hash of the proof will be the corresponding unencoded data commitment for the current layer  (either `CommR[l]` or `CommRLast`), (2) the leaf value is `ParentValue[l][c][p]`
+
+- `DataValue : [LAYERS][]Fr`: Value of the unencoded challenged nodes at layer `l`.
+
+- `ReplicaValue : [LAYERS][]Fr`: Value of the encoded nodes for each challenged node at layer `l`.
+
+- `ParentValue : [LAYERS][][PARENT_NODES]Fr`: Value of the parent nodes for each challenged node at layer `l`.
+
+##### Design notes
+
+#### Circuit
+
+##### High Level
+
+In high level, we do 4 checks:
+
+1. **ReplicaId Check**: Check the binary representation of the ReplicaId
+2. **Inclusion Proofs Checks**: Check the inclusion proofs
+3. **Encoding Checks**: Check that the data has been correctly encoding into a replica
+4. **CommRStar Check**: Check that CommRStar has been generated correctly
+
+##### Details
+
+```go
+// 1: ReplicaId Check - Check ReplicaId is equal to its bit representation
+let ReplicaIdBits : [255]Fr = Fr_to_bits(ReplicaId)
+assert(Packed(replica_id_bits) == ReplicaId)
+
+let DataRoot, ReplicaRoot Fr
+
+for l in range LAYERS {
+  
+  if l == 0 {
+    DataRoot = CommD
+  } else {
+    DataRoot = CommR[l-1]
+  }
+
+  if l == LAYERS-1 {
+    ReplicaRoot = CommRLast
+  } else {
+    ReplicaRoot = CommR[l]
+  }
+  
+  for c in range LAYERS_CHALLENGES[l] {
+    // 2: Inclusion Proofs Checks
+    // 2.1: Check inclusion proofs for data nodes are correct
+    assert(MerkleTreeVerify(DataRoot, InclusionPath[l][c], DataProof[l][c], DataValue[l][c]))
```

Comment:
let's define `MerkleTreeVerify` or link it to the high level check it's in charge of
(edit: I realize now there's the number linking it, so this is fine as is)